### PR TITLE
Hello World is 100% Safe

### PR DESCRIPTION
**Pull request by Lokathor**

### CODE REVIEW

File: book/src/04-non-video/01-buttons.md [hunk 1]
```diff
@@ -3,3 +3,81 @@
...
+the whole frame. If you've worked with polling input before that should sound
+totally normal, but if not just always remember to gather the input once per
+frame and then use that value across the whole frame.
+
+## Key Interrupt Control
+
+* KEYCNT, `0x400_0132`, `u16`, read/write
+
+This lets you control what keys will trigger a keypad interrupt. Of course, for
+the actual interrupt to fire you also need to set the `IME` and IE` registers
```
Comment (dannyfritz) [inline book/src/04-non-video/01-buttons.md]: `IE`

Comment (Lokathor) [reply book/src/04-non-video/01-buttons.md]: good catch

File: book/src/04-non-video/01-buttons.md [hunk 2]
```diff
@@ -3,3 +3,98 @@
...
+* Right
+* Left
+* Up
+* Down
+* R
+* L
+
+Bits above that are not used. However, since the arrow left and right can never
+be pressed at the same time, and the arrows up and down can never be pressed at
+the same time, this register will never read as zero.
```
Comment (dannyfritz) [inline book/src/04-non-video/01-buttons.md]: Which register can't read as 0?

File: book/src/04-non-video/01-buttons.md [hunk 3]
```diff
@@ -3,3 +3,98 @@
...
+  }
+}
+```
+
+Note that the current `KEYINPUT` value changes in real time as the user presses
+or releases the buttons. To account for this, it's best to read the value just
+once per game frame and then use that single value as if it was the input across
+the whole frame. If you've worked with polling input before that should sound
+totally normal, but if not just always remember to gather the input once per
+frame and then use that value across the whole frame.
```
Comment (dannyfritz) [inline book/src/04-non-video/01-buttons.md]: If you've worked with polling input before that should sound
totally __normal. But if not, just__ always remember to gather the input once per
frame and then use that value across the whole frame.

File: book/src/04-non-video/01-buttons.md (hunk 2 above)
Comment (Lokathor) [reply book/src/04-non-video/01-buttons.md]: `KEYINPUT`, edited for clarity

File: book/src/04-non-video/01-buttons.md (hunk 3 above)
Comment (Lokathor) [reply book/src/04-non-video/01-buttons.md]: Yeah, I like this

File: book/src/04-non-video/01-buttons.md [hunk 4]
```diff
@@ -3,3 +3,100 @@
...
+```
+
+Note that the current `KEYINPUT` value changes in real time as the user presses
+or releases the buttons. To account for this, it's best to read the value just
+once per game frame and then use that single value as if it was the input across
+the whole frame. If you've worked with polling input before that should sound
+totally normal. If not, just remember to call `read_key_input` once per frame
+and then use that `KeyInput` value across the whole frame.
+
+### Detecting New Presses
```
Comment (dannyfritz) [inline book/src/04-non-video/01-buttons.md]: Debouncing would be a good topic to talk about here. I'm assuming the hardware doesn't debounce for the software.

Comment (Lokathor) [reply book/src/04-non-video/01-buttons.md]: During an particular CPU cycle the buttons will be either on or off, but there's no more special processing than that.

**Pull request merged**